app/components/DataInput/DateFormatter/index.js

```diff
@@ -0,0 +1,107 @@
+import React, { Component } from 'react';
+import ReactDOM from 'react-dom';
+import { connect } from 'react-redux';
+import * as styles from '../DataInput.css';
+import { Switch, Label } from 'rebass';
+import DispatchField from '../../lib/DispatchField';
+import HelpTrigger from '../../lib/HelpTrigger';
+import { RECEIVE_DATE_FORMAT } from '../../../constants';
+import * as dateUtils from '../../../utils/parseDate';
+import { Text } from 'rebass';
+import actionTrigger from '../../../actions';
+
+class DateFormatter extends Component {
+  constructor() {
+    super();
+    this._toggleFormatter = this._toggleFormatter.bind(this);
+    this._handleChange = this._handleChange.bind(this);
+    this._validate = this._validate.bind(this);
+  }
+
+  componentWillReceiveProps(nextProps) {
+    if (nextProps.dates.toString() !== this.props.dates.toString()) {
+      this._validate(nextProps.dateFormat.formatString, nextProps.dates);
+    }
+  }
+
+  componentDidUpdate() {
+    if (this.props.dateFormat.enabled) {
+      ReactDOM.findDOMNode(this)
+        .querySelector('[name="formatString"]')
+        .focus();
+    }
+  }
+
+  _toggleFormatter(evt) {
+    evt.preventDefault();
+    this.props.dispatch(actionTrigger(RECEIVE_DATE_FORMAT, {
+      enabled: !this.props.dateFormat.enabled,
+    }));
+  }
+
+  _handleChange(fieldProps, newValue) {
+    this._validate(newValue, this.props.dates);
+  }
+
+  _validate(formatString, dates) {
+    const testResult = dateUtils.disproveList(formatString, dates);
+    this.props.dispatch(actionTrigger(RECEIVE_DATE_FORMAT, {
+      validated: !testResult,
+      failedAt: testResult,
+      formatString,
+    }));
+  }
+
+  render() {
```

Comment:
This is a REALLY large render, can it be broken up into multiple functions (or components)?